// openclaw/dist/outbound-adapter-BKJG6LOz.js
import { a as shouldLogVerbose } from "./globals-CX0Bl8ZU.js";
import { a as chunkText } from "./chunk-D7yuCiqB.js";
import "./runtime-env-DxZvavA4.js";
import "./reply-chunking-DwKx7qFI.js";
import { t as resolveWhatsAppOutboundTarget } from "./resolve-outbound-target-C5tZvHCQ.js";
import { n as normalizeWhatsAppPayloadText } from "./outbound-media-contract-DOU0vFqw.js";
import { t as createWhatsAppOutboundBase } from "./outbound-base-BoY6aVpE.js";
//#region extensions/whatsapp/src/outbound-adapter.ts
let whatsAppSendModulePromise;
function loadWhatsAppSendModule() {
	whatsAppSendModulePromise ??= import("./send-BLG3BiWl.js");
	return whatsAppSendModulePromise;
}
function normalizeOutboundText(text) {
	return normalizeWhatsAppPayloadText(text);
}
const whatsappOutbound = createWhatsAppOutboundBase({
	chunker: chunkText,
	sendMessageWhatsApp: async (to, text, options) => await (await loadWhatsAppSendModule()).sendMessageWhatsApp(to, normalizeOutboundText(text), { ...options }),
	sendPollWhatsApp: async (to, poll, options) => await (await loadWhatsAppSendModule()).sendPollWhatsApp(to, poll, options),
	shouldLogVerbose: () => shouldLogVerbose(),
	resolveTarget: ({ to, allowFrom, mode }) => resolveWhatsAppOutboundTarget({
		to,
		allowFrom,
		mode
	}),
	normalizeText: normalizeOutboundText,
	skipEmptyText: true
});
//#endregion
export { whatsappOutbound as t };
