// openclaw/dist/slash-dispatch.runtime-tFeMFFfZ.js
import { t as resolveConversationLabel$1 } from "./conversation-label-uLkSm7_X.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-BsDft3qL.js";
import { s as resolveChunkMode$1 } from "./chunk-D7yuCiqB.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-uGoWkLGV.js";
import "./routing-By0dru2e.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-l9GBKp8I.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-bLmSGU0b.js";
import "./reply-runtime-qSTWVO86.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-O9M0wjdI.js";
import "./markdown-table-runtime-BEgGw_S1.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-Bj61HHmX.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
